### ISSUE TITLE
btrfs on GPT with BIOS fails

### ISSUE THREAD

**Issue by wxl** [calamares/calamares#1014]
**Steps to reproduce:**
 1. Using BIOS, load Calamares up and make appropriate choices until the Partitions section
 2. Select manual partitioning and hit Next
 3. Click New Partition Table, select GPT, and Click ok
 4. Select Free Space and click Create
 5. Select the following and hit ok:
    1. File System: btrfs
    2. Mount Point: /
    3. Flags: boot
 6. Proceed with the installation

**Expected results:** successful installation.

**Actual results:** failed installation with the following error (full session.log is [here][1]):
```
Installation failed: 

Boost.Python error in job "bootloader".

Command 'grub-install --target=i386-pc --recheck --force /dev/sda' returned non-zero exit status 1.
Installing for i386-pc platform.grub-install: warning: this GPT partition label contains no BIOS Boot Partition; embedding won't be possible.grub-install: error: filesystem `btrfs' doesn't support blocklists.

Traceback:

File "/usr/lib/x86_64-linux-gnu/calamares/modules/bootloader/main.py", line 420, in run
     prepare_bootloader(fw_type)

  File "/usr/lib/x86_64-linux-gnu/calamares/modules/bootloader/main.py", line 387, in prepare_bootloader
    install_grub(efi_directory, fw_type)

  File "/usr/lib/x86_64-linux-gnu/calamares/modules/bootloader/main.py", line 304, in install_grub
    boot_loader["installPath"]])

  File "<string>", line 6, in <module> 
```

**Other notes:**
These all successfully install (using the same steps as above, except regarding partition type and file system type):
 * btrfs on MBR
 * ext4 on GPT
 * ext on MBR
 * btrfs on GPT with EFI

Explicitly selecting to install the bootloader on the System Partition resolves the issue (the installation succeeds), but then the installation is unbootable.

**Details:**
 * OS: Lubuntu 18.10 daily (20180814)
 * Calamares version: 3.2.1-1

[1]: https://share.riseup.net/#RGlp2nCXhXUnUZQvynAmbA

**Comment by abucodonosor** [comment]
Well the error is teling you what is going on .. and I doubt the way you describe is working for any 
FS as GPT on BIOS boot

For GPT on BIOS boot you need the first partition 8MB unformatted and with ' bios_grub ' flag set.

As always , while something like this may work , manual partitioning just means you 'know what your are doing' .. Calamares itself does not care about what flags you set or not set there.

**Comment by wxl** [comment]
…except for the fact that ext4 on GPT works just fine without changing
anything else. All on a single partition.

**Comment by kkofler** [comment]
The error message also says why this works on ext4 and not on btrfs: ext4 supports the "blocklist" hack, where GRUB locks the file to a fixed position in the file system and then refers to the blocks directly (i.e., the old LILO approach, with the drawback that you have to reinstall the bootloader to the MBR if you touch its files in the partition in any way), btrfs does not (I guess it does not support locking files to a fixed position).

So I think this is not a bug, it is just a technical limitation of GRUB that we have to deal with. However, there should probably be a warning or error about this from the partitioning UI so that you can fix it before proceeding to the actual installation (which is why I am keeping this issue open).

**Comment by mirh** [comment]
> ext4 on GPT

[Uhh?](459)

> btrfs on GPT with EFI

Oh, so your system is just efi with csm enabled...

Anyway AFAIU, [proper behavior](https://bugzilla.redhat.com/show_bug.cgi?id=748071) for GPT disks under bios assumptions, would be to have a [bios boot partition](https://en.wikipedia.org/wiki/BIOS_boot_partition).

**Comment by kkofler** [comment]
@abucodonosor: Sorry, but I think your reply contradicts the text of the error message from GRUB, and also previous user reports (as far as I can remember them):

> BTRFS works fine on BIOS GPT not a FS limitation.
>
> You'll see this error on any FS.

The missing BIOS boot partition can be worked around for some file systems such as ext4 (but not btrfs) using the "blocklists" hack. That setup is **not** recommended by GRUB upstream, but if you pass `--force` to GRUB, it will do it. This is exactly why Calamares always passes `--force` to GRUB.

So where did you get the information from that it does not work? Have you ever tried this with ext4? Or have you heard or read from anybody who did?

**Comment by abucodonosor** [comment]
@kkofler 

wth are you talking about ? 

We are talking about BIOS boot with 'GPT' ..

Ofc you can .. and it does work , ext* , XFS , BTRS ... 

Using "Blocklist" hack with force will result in a half-broken partition , if you can boot it ( by luck )
each time you run grub update your UUID on the partition breaks .. As a note try that without force 
and it will error out even on ext*.

The bug here is simple .. UIX won't set the flag . once set you can install and boot BIOS
with GPT as usual and the used FS does not matter.

>The missing BIOS boot partition can be worked around for some file systems such as ext4 (but not btrfs)

No wrong the other way around .. the "Blocklist" hack works ( well silent ingores the errors and break on runtime ) for ext* ..

To have working GPT on BIOS boot you need your first partition unformated at least 8MB and with bios boot flag set .. and that is all about.

**Comment by abucodonosor** [comment]
> This is exactly why Calamares always passes --force to GRUB.

And that is wrong and should be fixed too

**Comment by kkofler** [comment]
> Ofc you can .. and it does work , ext* , XFS , BTRS ...
>
> Using "Blocklist" hack with force will result in a half-broken partition , if you can boot it ( by luck )
> each time you run grub update your UUID on the partition breaks .. 

That's what I had written ("with the drawback that you have to reinstall the bootloader to the MBR if you touch its files in the partition in any way").

So we actually agree about the facts, just not about the wording.

> > This is exactly why Calamares always passes --force to GRUB.
>
> And that is wrong and should be fixed too

And how is it better to have the whole installation fail with an error than to do what the user explicitly asked for during manual partitioning?

Printing a warning or even an error **beforehand**, during partitioning, while the user can fix it, is OK. Failing the installation not so much when it can be avoided.

And if the flag is not getting set when the user checked the box in the UI, that's definitely a bug that needs to be fixed. I thought any such bugs have already been addressed, but I may be wrong.

**Comment by abucodonosor** [comment]
> And how is it better to have the whole installation fail with an error than to do what the user explicitly asked for during manual partitioning?

Yes much better.. The user really asked something diferent in this case .. and while we force
we hide the fact that the installation may break on runtime or it installs but does not boot ,
which is much much worse.

So just error out , the message about is clear and should tell everyone what is missing.

However this is not the only sitution we got other  bugs bc ext* blocklists" hacks .. Can't find the bug number yet but is about having ext* MBR and then trying to install grub again into a partition with ext* .. the result
is similar to that .

>Printing a warning or even an error beforehand, during partitioning, while the user can fix it, is OK. Failing the installation not so much when it can be avoided.

Sure but we don't really do anything with manual partitioning.. is up2 the user to know what to do.
At least the sitution now.

We do not support such things with auto so nothing is checked at all for manual partitioning.
Installation wil fail when user forgets to set boot flag , add wrong mount point etc .. we do not check that.

 Some 'user' errors on manual partitioning could emit a warning , I'm all for that.

( I think @demmm  has an UIX warning for the BIOS on GPT error in her tree ) 

We can even 'drop' using MBR and switch everyting to GPT as example ( right way ofc ).

> I thought any such bugs have already been addressed, but I may be wrong.

I can remeber @adriaandegroot fixed some of these too , yes.

Is why I'm not sure is an calamares bug or something in Ubuntu tools/kpmcore version

I may find some time and build a iso here with calamares 3.2.x ..
 I'm still on 3.1.* which does not have this bug in my setup.

**Comment by mirh** [comment]
> We can even 'drop' using MBR and switch everyting to GPT as example ( right way ofc ).

That would really be nice, but unfortunately there are a lot of [ifs and buts](https://www.rodsbooks.com/gdisk/booting.html) with bios and gpt (from bugs in the firmware, to.. well, _Windows_)

**Comment by abucodonosor** [comment]
@mirh 

Yes I know but I've talked about Linux installations only..

That can be a option in UIX too ofc so user can decide by itself without the need to use
manual partitioning.

**Comment by mirh** [comment]
Of course it should 100% totally be an option even for users that don't want to dirty their hands

But I was saying that there will be the need to have warnings for that (and ***unfortunately*** I can hardly see how those would play out, if having it default)

**Comment by Conan-Kudo** [comment]
@abucodonosor So it seems like the issue here is that BIOS+GPT _should not be offered_ unless Calamares is fixed to create a BIOS boot partition for GRUB, which is what other installers do when you do that combination. For example, if you boot Fedora live media on BIOS with the `inst.gpt` flag, Anaconda will install with GPT and create a BIOS boot partition for GRUB correctly.
